# DA0031 Analysing bulk RNA sequencing data zebra fish embryos
# DA0031-2021_Sep_01-analysis.1



## make directories for the tar file
```sh
mkdir -p /nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1/raw 
cd /nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1/raw

wget -c --no-check-certificate --auth-no-challenge --user 'jeanpierre.elbers' --ask-password https://ngs.vbcf.ac.at/filemanager/byurl/Kji01dY8of-HJHMFDSX2_4_R12083_20210828.tar.gz
```
## check the md5sum
```sh
cat Kji01dY8of-HJHMFDSX2_4_R12083_20210828.tar.gz.md5sum.txt
f5f0b8ffed3bcb8bcc5ef5339155d400 Kji01dY8of-HJHMFDSX2_4_R12083_20210828.tar.gz

md5sum --check Kji01dY8of-HJHMFDSX2_4_R12083_20210828.tar.gz.md5sum.txt > Kji01dY8of-HJHMFDSX2_4_R12083_20210828.tar.gz.md5sum.check 2>&1 &

cat Kji01dY8of-HJHMFDSX2_4_R12083_20210828.tar.gz.md5sum.check
#Kji01dY8of-HJHMFDSX2_4_R12083_20210828.tar.gz: OK


# untar

tar xzf Kji01dY8of-HJHMFDSX2_4_R12083_20210828.tar.gz > Kji01dY8of-HJHMFDSX2_4_R12083_20210828.tar.gz.untar.log 2>&1 &
```




## Get zebra fish reference
```sh
    mkdir -p /nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1/ref
    cd /nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1/ref
    wget http://ftp.ensembl.org/pub/release-104/fasta/danio_rerio/dna/Danio_rerio.GRCz11.dna.primary_assembly.fa.gz
    wget http://ftp.ensembl.org/pub/release-104/fasta/danio_rerio/cdna/Danio_rerio.GRCz11.cdna.all.fa.gz
```
## Make Salmon/1.1.0 reference
```sh
    module load salmon/1.1.0
    grep "^>" <(gunzip -c Danio_rerio.GRCz11.dna.primary_assembly.fa.gz) | cut -d " " -f 1 > decoys.txt
    sed -i.bak -e 's/>//g' decoys.txt
    cat Danio_rerio.GRCz11.cdna.all.fa.gz Danio_rerio.GRCz11.dna.primary_assembly.fa.gz > Danio_rerio.GRCz11_gentrome.fa.gz
    salmon index -t Danio_rerio.GRCz11_gentrome.fa.gz -d decoys.txt -p 20 -i Danio_rerio.GRCz11.Ensembl104.salmon.1.1.0.idx \
    2> Danio_rerio.GRCz11.Ensembl104.salmon.1.1.0.idx.stderr 1> Danio_rerio.GRCz11.Ensembl104.salmon.1.1.0.idx.stdout

    zcat Danio_rerio.GRCz11.cdna.all.fa.gz |grep ">" |cut -f 1,4 -d " "|perl -pe "s/>//g"|perl -pe "s/gene://g" |tr ' ' '\t' > Danio_rerio.GRCz11.Ensembl104.transcripts2genes.tsv
```

## convert transcripts to gene symbols for Enhanced volcano plot
```sh
zcat Danio_rerio.GRCz11.cdna.all.fa.gz|grep ">" |cut -f 1,7 -d " " |perl -pe "s/>//g" |tr ' ' '\t' |perl -pe "s/gene_symbol://g" > transcript2geneSymbol.txt
```

## convert Ensembl gene ids to gene symbols for Enhanced volcano plot
```sh
zcat Danio_rerio.GRCz11.cdna.all.fa.gz|grep ">" |cut -f 4,7 -d " " |perl -pe "s/gene://g" |tr ' ' '\t' |perl -pe "s/gene_symbol://g" > gene2geneSymbol.txt
```


samples.txt
```sh
168952	Animal_Ecto_21.6.21	RNAseq.Animal.1
168953	Animal_ecto_25.6.21	RNAseq.Animal.2
168954	Animal_ecto_5.7.21	RNAseq.Animal.3
168955	Animal_Ecto_9.7.21	RNAseq.Animal.4
168956	Animal_Ecto_13.7.21	RNAseq.Animal.5
168957	Lateral_Ecto_17.6.21	RNAseq.Lateral.1
168958	Lateral_Ecto_24.6.21	RNAseq.Lateral.2
168959	Lateral_Ecto_2.7.21	RNAseq.Lateral.3
168960	Lateral_ecto_7.7.21	RNAseq.Lateral.4
168961	Lateral_Ecto_12.7.21	RNAseq.Lateral.5
168962	Ctrl_7.7.21	RNAseq.Ctrl.1
168963	Ctrl_12.7.21	RNAseq.Ctrl.2
168964	Ctrl_13.7.21	RNAseq.Ctrl.3
168965	All_white_19.7.21	RNAseq.All.1
168966	All_Red_19.7.21	RNAseq.All.2
168967	All_Blue_19.7.21	RNAseq.All.3
```


## make symlinks for easier read identification
```sh
while read a b
do
  ln -s HJHMFDSX2_4_R12083_20210828/demultiplexed/${a}/${a}*R1*.fastq.gz ${b}.mate1.fastq.gz
  ln -s HJHMFDSX2_4_R12083_20210828/demultiplexed/${a}/${a}*R2*.fastq.gz ${b}.mate2.fastq.gz
done < <(cut -f 1,3 samples.txt)
```













## prepare for Snakemake analysis pipeline for differential gene expression
```sh
cd /nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1
```

config.yaml
```sh
# defining the sample/file names
#	Experiment	... <string> without \s \.
#	Condition{1,}	... <string> without \s \.
#	rep{1,2,...,N}	... "rep" + \d+
#	files with {dataset}.{mate1,mate2}.fastq.gz or {dataset}.bam must be in the directory "raw"
datasets: [RNAseq.Animal.1,
           RNAseq.Animal.2,
           RNAseq.Animal.3,
           RNAseq.Animal.4,
           RNAseq.Animal.5,
           RNAseq.Lateral.1,
           RNAseq.Lateral.2,
           RNAseq.Lateral.3,
           RNAseq.Lateral.4,
           RNAseq.Lateral.5,
           RNAseq.Ctrl.1,
           RNAseq.Ctrl.2,
           RNAseq.Ctrl.3,
           RNAseq.All.1,
           RNAseq.All.2,
           RNAseq.All.3]

conditions: [Animal,
             Lateral,
             Ctrl,
             All]

# defining the number of threads used for each process
#	must be an <integer>
threads: "8"

# trimmomatic paramters
trimmomatic-params: "2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:19"
# path to a fasta file with the adapter sequences to be used for clipping
trimmomatic-fasta: "/nfs/scistore14/rcsb/pub/biogrp/ReferenceData/AdapterSequences/Adapter.fa"

# salmon parameters
# path to salmon index directory
salmon-idx: "/nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1/ref/Danio_rerio.GRCz11.Ensembl104.salmon.1.1.0.idx"
# path to file holding the mapping between transcript IDs and gene ID
salmon-t2g: "/nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1/ref/Danio_rerio.GRCz11.Ensembl104.transcripts2genes.tsv"
# library type:
#	"A"	...  Automatic library type detection (see https://salmon.readthedocs.io/en/latest/salmon.html#what-s-this-libtype)
salmon-libtype: "A"
# additional paramters for salmon forwarded to the salmon call
salmon-params: " --seqBias --gcBias "

# dgea parameters
# specifiy comparisons to be made: <condition1-condition2::condition1-condition3:: ... >
dgea-cmps: "Animal-Lateral"
```












id2sample.txt
```sh
Animal.1        Animal_Ecto_21.6.21
Animal.2        Animal_ecto_25.6.21
Animal.3        Animal_ecto_5.7.21
Animal.4        Animal_Ecto_9.7.21
Animal.5        Animal_Ecto_13.7.21
Lateral.1       Lateral_Ecto_17.6.21
Lateral.2       Lateral_Ecto_24.6.21
Lateral.3       Lateral_Ecto_2.7.21
Lateral.4       Lateral_ecto_7.7.21
Lateral.5       Lateral_Ecto_12.7.21
```



















Snakefile
```sh
# specify name of external config-file
configfile: "config.yaml"

# import custom global parameter from config-file
DATASETS = config["datasets"]
CONDITIONS = config["conditions"]
THREADS = config["threads"]
dgeaCmpsString = config["dgea-cmps"]
dgeaCmpsList = dgeaCmpsString.split("::")

# list of rules which are not deployed to slurm
localrules: all, fastqc, quant, dgea, clean, salmon_index, trimmomatic_fasta, deseq2_dgea


# final target rule to produce all sub targets
rule all:
    input:
        fastqc = "logs/fastqc.done",
        quant  = "logs/quant.done",
        dgea   = "logs/dgea.done"

# subtarget: bam to fastq conversion
rule fastqc:
    input: 
        fastq1 = expand("qc/fastqc_trimmed/{dataset}-trimmed.mate1_fastqc.html", dataset=DATASETS),
        fastq2 = expand("qc/fastqc_trimmed/{dataset}-trimmed.mate2_fastqc.html", dataset=DATASETS),
        fastq3 = expand("qc/fastqc_raw/{dataset}.mate1_fastqc.html", dataset=DATASETS),
        fastq4 = expand("qc/fastqc_raw/{dataset}.mate2_fastqc.html", dataset=DATASETS)
    output: "logs/fastqc.done"
    shell:
        '''
        touch {output}
        '''

# subtarget: read per transcript quantification
rule quant:
    input: 
        bam = expand("quants/{dataset}/alignments.bam", dataset=DATASETS)
    output: "logs/quant.done"
    shell:
        '''
        touch {output}
        '''

# subtarget: do differential gene expression analysis
rule dgea:
    input: 
        csv = expand("dgea/DESeq2_{cmp}.csv", cmp = dgeaCmpsList)
    output: "logs/dgea.done"
    shell:
        '''
        touch {output}
        '''


###
######
###

# step 2: check read quality of raw reads
rule fastqc_check:
    input:
        mate1 = "raw/{sample}.mate1.fastq.gz",
        mate2 = "raw/{sample}.mate2.fastq.gz"
    output:
        mate1 = "qc/fastqc_raw/{sample}.mate1_fastqc.html",
        mate2 = "qc/fastqc_raw/{sample}.mate2_fastqc.html"
    params:
        threads = THREADS,
        dir = "qc/fastqc_raw"
    shell:
        '''
        module load fastqc/0.11.7
        fastqc --outdir {params.dir} --thread {params.threads} {input.mate1} {input.mate2}  > /dev/null 2> /dev/null
        '''


# step 4: clip/trimm reads with trimmomatic
rule trimmomatic_fasta:
    input: config["trimmomatic-fasta"]
    output: "auxData/trimmomatic-fasta.fa"
    shell:
        '''
        cp {input} {output}
        '''

rule trimmomatic:
    input:
        mate1 = "raw/{sample}.mate1.fastq.gz",
        mate2 = "raw/{sample}.mate2.fastq.gz",
        adapter = "auxData/trimmomatic-fasta.fa"
    output:
        mate1 = "fastq/{sample}-trimmed.mate1.fastq.gz",
        mate2 = "fastq/{sample}-trimmed.mate2.fastq.gz",
        unpaired1 = "fastq/{sample}.unpairedMate1.fastq.gz",
        unpaired2 = "fastq/{sample}.unpairedMate2.fastq.gz",
    log:
        err = "logs/trimmomatic/{sample}.log"
    params:
        threads = THREADS,
        params = config["trimmomatic-params"]       
    shell:
        '''
        unset SLURM_EXPORT_ENV
        module load trimmomatic/0.38
        java -jar /mnt/nfs/clustersw/shared/trimmomatic/0.38/trimmomatic-0.38.jar PE -threads {params.threads} -phred33 {input.mate1} {input.mate2} {output.mate1} {output.unpaired1} {output.mate2} {output.unpaired2} ILLUMINACLIP:{input.adapter}:{params.params}  > /dev/null 2> {log.err}
        '''

# step 5: check read quality of trimmed reads
rule fastqc_recheck:
    input:
        mate1 = "fastq/{sample}-trimmed.mate1.fastq.gz",
        mate2 = "fastq/{sample}-trimmed.mate2.fastq.gz"
    output:
        mate1 = "qc/fastqc_trimmed/{sample}-trimmed.mate1_fastqc.html",
        mate2 = "qc/fastqc_trimmed/{sample}-trimmed.mate2_fastqc.html"
    params:
        threads = THREADS
    shell:
        '''
        module load fastqc/0.11.7
        fastqc --outdir qc/fastqc_trimmed/ --thread {params.threads} {input.mate1} {input.mate2}  > /dev/null 2> /dev/null
        '''


# step 6: link to prepared salmon index (v1.1.0)
rule salmon_index:
    input:
        idx = config["salmon-idx"],
        t2g = config["salmon-t2g"]
    output:
        idx = directory("auxData/salmon.idx"),
        t2g = "auxData/t2g.csv"
    shell:
        '''
        cp -r {input.idx} {output.idx}
        cp {input.t2g} {output.t2g}
        '''


# step 7: quantify tanscript abundance with salmon (v1.1.0)
rule salmon_quant:
    input:
        mate1 = "fastq/{sample}-trimmed.mate1.fastq.gz",
        mate2 = "fastq/{sample}-trimmed.mate2.fastq.gz",
        idx   = "auxData/salmon.idx",
        t2g   = "auxData/t2g.csv"
    output:
        sf  = "quants/{sample}/quant.sf",
        gsf = "quants/{sample}/quant.genes.sf",
        sam = temp("quants/{sample}/alignments.sam"),
    log:
        err = "logs/salmon_quant/{sample}.err",
        out = "logs/salmon_quant/{sample}.out"
    params:
        dir = "quants/{sample}",
        threads = THREADS,
        params = config["salmon-params"],
        libtype = config["salmon-libtype"]
    shell:
        '''
        module load salmon/1.1.0
        salmon quant --quiet --threads {params.threads} {params.params} -i {input.idx} --libType {params.libtype} -1 {input.mate1} -2 {input.mate2} --geneMap {input.t2g} --writeMappings={output.sam} --output {params.dir}  > {log.out} 2> {log.err} 
        '''

# step 8: generate sorted bam file from salmon generated sam file
rule sam2sortedbam:
    input: "quants/{sample}/alignments.sam"
    output: "quants/{sample}/alignments.bam"
    log: "logs/sam2sortedbam/{sample}.log"
    params:
        threads = THREADS
    shell:
        '''
        module load samtools/1.10
        samtools view -@{params.threads} -b {input} | samtools sort -@{params.threads} - -o {output} 2> {log}
        '''

# step 9: make differential gene experession analysis
# run external R-script (scripts/DESeq2.R)
# see script for requirments concerning file name pattern
rule deseq2_dgea:
    input:
        quants = expand("quants/{dataset}/quant.sf", dataset=DATASETS),
        g2t = "auxData/t2g.csv",
        bin = "scripts/DESeq2_multiComp.R"
    output:
        pca  = "dgea/DESeq2_pca.pdf",
        heat = "dgea/DESeq2_heatmap.pdf",
        csv  = expand("dgea/DESeq2_{cmp}.csv", cmp = dgeaCmpsList),
        cnts = "dgea/DESeq2_counts.csv",
        tpm  = expand("dgea/DESeq2_TPM_{cond}.csv", cond = CONDITIONS),
        cnt  = expand("dgea/DESeq2_counts_{cmp}.csv", cmp = dgeaCmpsList),
        ma   = expand("dgea/DESeq2_MAplot_{cmp}.pdf", cmp = dgeaCmpsList),
        si   = "dgea/RsessionInfo.txt"
    params:
        cmps = dgeaCmpsString
    log:
        err = "logs/deseq2_dgea/DESeq2.err",
        out = "logs/deseq2_dgea/DESeq2.out"
    shell:
        '''
        module purge
        module load R/4.0.3
        Rscript --no-save --no-restore {input.bin} {input.g2t} {params.cmps} {input.quants} > {log.out} 2> {log.err}
        '''

###
######
###

# rule to remove all workflow intermediate and results files, but the logs
rule clean:
    shell:
        '''
        rm -rf fastq/
        rm -rf quants/
        rm -rf qc/
        rm -rf dgea/
        rm -rf auxData/
        '''
```








scripts/DESeq2_multiComp.R 
```sh
# load packages
  library("tximport")
  library("DESeq2")
  library("RColorBrewer")
  library("pheatmap")
  library("readr")
  library("apeglm")
  library("plyr")
  library("genefilter")
  library("ggplot2")
  library("ggrepel")

# defined custom operators and functions
  `%notin%` <- Negate(`%in%`)
  
  colSelectMean <- function(df){
    returnTab <- data.frame(matrix(ncol = length(unique(colnames(df))), nrow = nrow(df)))
    colnames(returnTab) <- unique(colnames(df))
    rownames <- rownames(df)
  
    for (c in 1:length(colnames(returnTab))){
      cn <- colnames(returnTab)[c]
      returnTab[,c] <- rowMeans(df[,colnames(df) == cn])
    }   
    rownames(returnTab) <- rownames
    return(returnTab)
  }


# parsing command line arguments
# order matters:
#       1.      transcript to gene ID mapping file
#       2.      string specifying the comparison of interests: cond1-cond2::cond3-cond4
#       3..N    quant.sf files from salmon
  args       <- commandArgs(trailingOnly= TRUE)
  files      <- args[-c(1,2)]
  g2t        <- args[1]
  compstring <- args[2]

# get sample names from snakemake variable
#       samples names must follow the pattern <STRING>.<CONDITION>.<repN>
#       whereas each section must not contain a "." and N must be an integer
  samples <- unlist(lapply(files, FUN = function(egtstr){tail(strsplit(egtstr, "/")[[1]], n = 2)[1]}))
  samples <- gsub("^[^.]+\\.", "", samples)
  names(files) <- samples

# get gene to transcript map file
  tx2gene <- read.table(g2t)
  txi <- tximport(files, type="salmon", tx2gene = tx2gene)
# get comparisons of interest from command line argument
  compOI      <- unlist(strsplit(compstring, split = "::"))

# remove .rep\d to get from sample name to condition
  conditions <- gsub(".\\d+$", "", samples, perl = TRUE)

# construct colData from samples and conditions
  colData <- data.frame(sample = samples, condition = factor(conditions))
 rownames(colData) <- samples

# run DESeq2
  ddsTxi <- DESeqDataSetFromTximport(txi,
                                   colData = colData,
                                   design = ~ condition)
  ddsTxi <- DESeq(ddsTxi)

# extract comparisons of interest
  uniqCond         <- unique(conditions)
  for (i1 in 1:length(uniqCond)){
    for (i2 in 1:length(uniqCond)){
      if(i1==i2){
        next;
      }

      cond1 <- uniqCond[i1]
      cond2 <- uniqCond[i2]

      comparison <- paste(cond1,cond2, sep = "-")

      if(any(comparison %notin% compOI)){
        next;
      }

      print(paste0("### start making contrast: ", cond1, " vs. ", cond2))

      res <- results(ddsTxi, contrast=c("condition", cond1, cond2))
      resLFC <- lfcShrink(ddsTxi, contrast=c("condition", cond1, cond2), type = "ashr")

# write result table to file
      res <- as.data.frame(res[order(res$pvalue),])
      res <- cbind(ID = rownames(res), res)
      rownames(res) <- NULL
      filename1 <- paste0("dgea/DESeq2_", comparison, ".csv")
      write.table(res, file=filename1, quote = F, sep="\t", row.names = F)

# compile table with raw, normalized, and mean normalized counts for each pair of compared samples and write to file

      counts.df     <- counts(ddsTxi,normalized=FALSE)
      normCounts.df <- counts(ddsTxi,normalized=TRUE)
      tpm.df        <- txi$abundance

      samplesOI     <- samples[conditions %in% cond1 | conditions %in% cond2]
      conditionsOI  <- conditions[conditions %in% cond1 | conditions %in% cond2]

      counts.df     <- counts.df[,colnames(counts.df) %in% samplesOI]
      normCounts.df <- normCounts.df[,colnames(normCounts.df) %in% samplesOI]
      tpm.df        <- tpm.df[,colnames(tpm.df) %in% samplesOI]

      colnames(normCounts.df) <- conditionsOI
      colnames(tpm.df)        <- conditionsOI
      meanNormCounts.df       <- colSelectMean(normCounts.df)
      meanTpm.df              <- colSelectMean(tpm.df)

      colnames(counts.df)         <- paste0("raw.", samplesOI)
      colnames(normCounts.df)     <- paste0("norm.", samplesOI)
      colnames(tpm.df)            <- paste0("TPM.", samplesOI)
      colnames(meanNormCounts.df) <- paste0("meanNorm.", colnames(meanNormCounts.df))
      colnames(meanTpm.df)        <- paste0("meanTPM.", colnames(meanTpm.df))

      joinTables.list <- list( counts=as.data.frame(counts.df), normCounts=as.data.frame(normCounts.df), meanNormCounts=as.data.frame(meanNormCounts.df), tpm=as.data.frame(tpm.df), meanTPM=as.data.frame(meanTpm.df) )
      for(i in 1:length(joinTables.list)){
        joinTables.list[[i]]$ID  <- rownames(joinTables.list[[i]])
      }

      fullCountTable <- join_all( joinTables.list, by="ID", type="full" )
      rownames(fullCountTable) <- fullCountTable$ID
      fullCountTable$ID <- NULL
      fullCountTable <- cbind(rownames(fullCountTable), fullCountTable)
      colnames(fullCountTable)[1] <- "ID"

      filename2 <- paste0("dgea/DESeq2_counts_", comparison, ".csv")
      write.table(as.data.frame(fullCountTable), file=filename2, quote = F, sep = "\t", row.names = FALSE)


# create MA plot
      filename3 <- paste0("dgea/DESeq2_MAplot_", comparison, ".pdf")
      pdf(file=filename3)
      plotMA(resLFC)
      dev.off()

      rm(cond1, cond2, res, resLFC, filename1, filename2, filename3, counts.df, normCounts.df, tpm.df, samplesOI, conditionsOI, joinTables.list, fullCountTable)
    }
  }

# make heat map of sample similarity
  vsdBM <- varianceStabilizingTransformation(ddsTxi, blind=FALSE)
  sampleDists <- dist(t(assay(vsdBM)))
  sampleDistMatrix <- as.matrix(sampleDists)
  rownames(sampleDistMatrix) <- vsdBM$sample
  colnames(sampleDistMatrix) <- NULL
  colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
  p<-pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors,
         filename = "dgea/DESeq2_heatmap.pdf")

# make PCA plot of sample similarity
pca_data <- DESeq2::plotPCA(vsdBM, intgroup=c("condition", "sample"), returnData = TRUE)
id2sample <- read.table("id2sample.txt")

pca_data2 <- merge(pca_data,id2sample,by.x="sample",by.y="V1")
p <- ggplot(data = pca_data2, aes(x = PC1, y = PC2)) + theme_bw() + geom_point(aes(col = condition), size = 5, alpha = 0.75) + geom_text_repel(aes(label = V2), size = 2.5)
ggsave(filename="dgea/DA0031_2021_Sep_01_analysis.1.pca.pdf", plot = p, width = 6.5, height = 6)


# write for each condition a file with all the samples TPM values and the mean and stddev of each gene
  tpm.df           <- txi$abundance

  for (currCond in uniqCond){
    currTpm.df  <- tpm.df[,grepl(paste0(currCond, "."), colnames(tpm.df))]
    writeTpm.df <- cbind(currTpm.df, rowMeans(currTpm.df), rowSds(currTpm.df))
    colnames(writeTpm.df)[ncol(writeTpm.df)]   <- "Stddev"
    colnames(writeTpm.df)[ncol(writeTpm.df)-1] <- "Mean"
    currFile = paste0("dgea/DESeq2_TPM_", currCond, ".csv")
    write.table(as.data.frame(writeTpm.df), file=currFile, quote = F, sep = "\t", row.names = TRUE)
  }

# compile table with raw, normalized, and mean normalized counts for all samples and write to file
  counts.df     <- counts(ddsTxi,normalized=FALSE)
  normCounts.df <- counts(ddsTxi,normalized=TRUE)
  tpm.df        <- txi$abundance

  colnames(normCounts.df) <- conditions
  colnames(tpm.df)        <- conditions
  meanNormCounts.df       <- colSelectMean(normCounts.df)
  meanTpm.df              <- colSelectMean(tpm.df)

  colnames(counts.df)         <- paste0("raw.", samples)
  colnames(normCounts.df)     <- paste0("norm.", samples)
  colnames(tpm.df)            <- paste0("TPM.", samples)
  colnames(meanNormCounts.df) <- paste0("meanNorm.", colnames(meanNormCounts.df))
  colnames(meanTpm.df)        <- paste0("meanTPM.", colnames(meanTpm.df))

  joinTables.list <- list( counts=as.data.frame(counts.df), normCounts=as.data.frame(normCounts.df), meanNormCounts=as.data.frame(meanNormCounts.df), tpm=as.data.frame(tpm.df), meanTPM=as.data.frame(meanTpm.df) )
  for(i in 1:length(joinTables.list)){
    joinTables.list[[i]]$ID  <- rownames(joinTables.list[[i]])
  }
 
  fullCountTable <- join_all( joinTables.list, by="ID", type="full" )
  rownames(fullCountTable) <- fullCountTable$ID
  fullCountTable$ID <- NULL
  fullCountTable <- cbind(rownames(fullCountTable), fullCountTable)
  colnames(fullCountTable)[1] <- "ID"

  write.table(as.data.frame(fullCountTable), file="dgea/DESeq2_counts.csv", quote = F, sep = "\t", row.names = FALSE)


# dump session info into file
  writeLines(capture.output(sessionInfo()), "dgea/RsessionInfo.txt")
 ```
 
 
 
 
 
 
 
 
cluster.json
```sh
{
    "__default__":
    {
        "account": "jelbers",
        "mem": "12G",
        "time": "24:00:0",
        "job-name" : "{rule}",
        "partition" : "defaultp"
    },
    "salmon_quant":
    {
        "account": "jelbers",
        "mem": "64G",
        "time": "24:00:0",
        "job-name" : "{rule}",
        "partition" : "defaultp"
    }
}
```
 
 

## set up Snakemake environment and run the analysis
```sh
conda activate snakemake-6.3.0
snakemake --snakefile Snakefile --dryrun  --dag all | dot -Tsvg > dag.svg
snakemake --snakefile Snakefile --printshellcmds --latency-wait 60 --local-cores 1 --cores all --cluster-config cluster.json --cluster "sbatch --export=NONE --no-requeue --job-name {cluster.job-name} --mem={cluster.mem} --time={cluster.time} --cores={THREADS} " all > snakemake.log 2>&1 &
```




## make multiqc report of before and after trimming
```sh
cd /nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1/qc

conda activate python-3.9.4

/nfs/scistore16/itgrp/jelbers/.local/bin/multiqc --interactive . --title DA0031-2021_Sep_01-analysis.1-trimming
```





## manully run the dgea analysis
```sh
cd /nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1/
mkdir -p dgea
module load R/4.0.3

# all samples
# saved to /home/jelbers/Downloads/heisenberg/DA0031/2021_Sep_01/analysis.1/dgea/all-animal-all-lateral
Rscript --no-save --no-restore scripts/DESeq2_multiComp.R auxData/t2g.csv Animal-Lateral quants/RNAseq.Animal.1/quant.sf quants/RNAseq.Animal.2/quant.sf quants/RNAseq.Animal.3/quant.sf quants/RNAseq.Animal.4/quant.sf quants/RNAseq.Animal.5/quant.sf quants/RNAseq.Lateral.1/quant.sf quants/RNAseq.Lateral.2/quant.sf quants/RNAseq.Lateral.3/quant.sf quants/RNAseq.Lateral.4/quant.sf quants/RNAseq.Lateral.5/quant.sf > logs/deseq2_dgea/DESeq2.out 2> logs/deseq2_dgea/DESeq2.err &


# minus two outliers
# saved to /home/jelbers/Downloads/heisenberg/DA0031/2021_Sep_01/analysis.1/dgea/minus-animal.1-lateral.4
Rscript --no-save --no-restore scripts/DESeq2_multiComp.R auxData/t2g.csv Animal-Lateral quants/RNAseq.Animal.2/quant.sf quants/RNAseq.Animal.3/quant.sf quants/RNAseq.Animal.4/quant.sf quants/RNAseq.Animal.5/quant.sf quants/RNAseq.Lateral.1/quant.sf quants/RNAseq.Lateral.2/quant.sf quants/RNAseq.Lateral.3/quant.sf quants/RNAseq.Lateral.5/quant.sf > logs/deseq2_dgea/DESeq2.out 2> logs/deseq2_dgea/DESeq2.err &
```

## on a local computer
```sh
cd /home/jelbers/Downloads/heisenberg/DA0031/2021_Sep_01/analysis.1/dgea
rsync --stats --progress --archive jelbers@bea81.ista.local://nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1/ref/transcript2geneSymbol.txt .
rsync --stats --progress --archive jelbers@bea81.ista.local://nfs/scistore16/itgrp/bioinf/projects/DA0031/2021_Sep_01/analysis.1/ref/gene2geneSymbol.txt .

cd /home/jelbers/Downloads/heisenberg/DA0031/2021_Sep_01/analysis.1/dgea/minus-animal.1-lateral.4

cp DESeq2_Animal-Lateral.csv DESeq2_Animal-Lateral_enhanced_volcano.csv

while read a b
do
 perl -pi -e "s/${a}/${b}/g" DESeq2_Animal-Lateral_enhanced_volcano.csv
done < ../gene2geneSymbol.txt &



cd /home/jelbers/Downloads/heisenberg/DA0031/2021_Sep_01/analysis.1/dgea/all-animal-all-lateral

cp DESeq2_Animal-Lateral.csv DESeq2_Animal-Lateral_enhanced_volcano.csv

while read a b
do
 perl -pi -e "s/${a}/${b}/g" DESeq2_Animal-Lateral_enhanced_volcano.csv
done < ../gene2geneSymbol.txt &
```

## activate correct conda environment for volcano plots
```sh
cd /home/jelbers/Downloads/heisenberg/DA0031/2021_Sep_01/analysis.1/dgea

conda activate enhancedvolcano-1.10.0
```

in R

```R
library(EnhancedVolcano)
library(biomaRt)
#install.packages("xml2")


# download biomart data
mart <- useDataset("drerio_gene_ensembl", useMart("ensembl"))

# read in comma-delimited DESeq output
minus2 <- read.table("minus-animal.1-lateral.4/DESeq2_Animal-Lateral.csv",header=T)

# get rid of ensembl gene id version
minus2$ENSEMBL <- gsub("\\.\\d+","",minus2$ID,perl=t)

# Get gene id and gene symbol
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","zfin_id_symbol"),values=minus2$ENSEMBL,mart= mart)
G_list$zfin_id_symbol <- gsub("^$",NA,G_list$zfin_id_symbol)

for (i in 1:length(G_list$zfin_id_symbol)) {
  if (is.na(G_list$zfin_id_symbol[i])) {
    G_list$zfin_id_symbol[i] <- G_list$ensembl_gene_id[i]
  }
}



df2 <- merge(minus2,G_list,by.x="ENSEMBL",by.y="ensembl_gene_id")

rownames(df2) <- make.names(df2$zfin_id_symbol, unique = TRUE)

pdf("enhanced-volcano-minus-animal.1-lateral.4.pdf", width=8.5, height=11)
  EnhancedVolcano(df2,
  lab = rownames(df2),
  x = 'log2FoldChange',
  y = 'pvalue',
  title = 'Animal versus Lateral DESeq2 minus two outliers',
  pCutoff = 1e-04,
  FCcutoff = 1,
  pointSize = 3.0,
  labSize = 6.0,
  labCol = 'black',
  labFace = 'bold',
  boxedLabels = TRUE,
  parseLabels = TRUE,
  col = c('black', 'pink', 'purple', 'red3'),
  colAlpha = 4/5,
  legendPosition = 'bottom',
  legendLabSize = 14,
  legendIconSize = 4.0,
  drawConnectors = TRUE,
  widthConnectors = 1.0,
  colConnectors = 'black') + coord_flip()
dev.off()

# read in comma-delimited DESeq output
minus2 <- read.table("all-animal-all-lateral/DESeq2_Animal-Lateral.csv",header=T)

# get rid of ensembl gene id version
minus2$ENSEMBL <- gsub("\\.\\d+","",minus2$ID,perl=t)

# Get gene id and gene symbol
G_list <- getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","zfin_id_symbol"),values=minus2$ENSEMBL,mart= mart)
G_list$zfin_id_symbol <- gsub("^$",NA,G_list$zfin_id_symbol)

for (i in 1:length(G_list$zfin_id_symbol)) {
  if (is.na(G_list$zfin_id_symbol[i])) {
    G_list$zfin_id_symbol[i] <- G_list$ensembl_gene_id[i]
  }
}



df2 <- merge(minus2,G_list,by.x="ENSEMBL",by.y="ensembl_gene_id")

rownames(df2) <- make.names(df2$zfin_id_symbol, unique = TRUE)

pdf("enhanced-volcano-all-animal-all-lateral.pdf", width=8.5, height=11)
  EnhancedVolcano(df2,
  lab = rownames(df2),
  x = 'log2FoldChange',
  y = 'pvalue',
  title = 'Animal versus Lateral DESeq2 all 5 bio replicates',
  pCutoff = 1e-04,
  FCcutoff = 1,
  pointSize = 3.0,
  labSize = 6.0,
  labCol = 'black',
  labFace = 'bold',
  boxedLabels = TRUE,
  parseLabels = TRUE,
  col = c('black', 'pink', 'purple', 'red3'),
  colAlpha = 4/5,
  legendPosition = 'bottom',
  legendLabSize = 14,
  legendIconSize = 4.0,
  drawConnectors = TRUE,
  widthConnectors = 1.0,
  colConnectors = 'black') + coord_flip()
dev.off()
quit()
```














## activate the next conda environment for functional enrichment plots

```sh
conda activate clustifyr-1.5.0
```

```R
library(clusterProfiler)
library(enrichplot)
# we use ggplot2 to add x axis labels (ex: ridgeplot)
library(ggplot2)


# SET THE DESIRED ORGANISM HERE
organism = "org.Dr.eg.db"
#BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)
library(biomaRt)
#install.packages("xml2")



df <- read.table("minus-animal.1-lateral.4/DESeq2_Animal-Lateral.csv",header=T)

# get rid of ensembl gene id version
df$ID <- gsub("\\.\\d+","",df$ID,perl=t)


df <- df[df$padj >= 0.05,]

# we want the log2 fold change 
original_gene_list <- df$log2FoldChange


# name the vector
names(original_gene_list) <- df$ID

# omit any NA values 
gene_list<-na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)

organism = org.Dr.eg.db
gse <- gseGO(
gene_list,
ont ="ALL", 
keyType = "ENSEMBL", 
OrgDb = organism, 
exponent = 1,
minGSSize = 10,
maxGSSize = 500,
eps = 1e-10,
pvalueCutoff = 0.05,
pAdjustMethod = "none",
verbose = TRUE,
seed = FALSE,
by = "fgsea")
dotplot(gse, showCategory=10, split=".sign",color="pvalue") + facet_grid(.~.sign)
write.table(x=gse@result,file="gene-set-enrichment-minus-animal.1-lateral.4.txt",eol="\n",row.names=F,col.names=T,quote=F,sep="\t")

pdf("gene-set-enrichment-minus-animal.1-lateral.4.pdf",height=11,width=8.5)
dotplot(gse, showCategory=10, split=".sign",color="pvalue") + facet_grid(.~.sign)
dev.off()




df <- read.table("all-animal-all-lateral/DESeq2_Animal-Lateral.csv",header=T)

# get rid of ensembl gene id version
df$ID <- gsub("\\.\\d+","",df$ID,perl=t)


df <- df[df$padj >= 0.05,]

# we want the log2 fold change 
original_gene_list <- df$log2FoldChange


# name the vector
names(original_gene_list) <- df$ID

# omit any NA values 
gene_list<-na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)

organism = org.Dr.eg.db
gse <- gseGO(
gene_list,
ont ="ALL", 
keyType = "ENSEMBL", 
OrgDb = organism, 
exponent = 1,
minGSSize = 10,
maxGSSize = 500,
eps = 1e-10,
pvalueCutoff = 0.05,
pAdjustMethod = "none",
verbose = TRUE,
seed = FALSE,
by = "fgsea")
dotplot(gse, showCategory=10, split=".sign",color="pvalue") + facet_grid(.~.sign)
write.table(x=gse@result,file="gene-set-enrichment-all-animal-all-lateral.txt",eol="\n",row.names=F,col.names=T,quote=F,sep="\t")


pdf("gene-set-enrichment-all-animal-all-lateral.pdf",height=11,width=8.5)
dotplot(gse, showCategory=10, split=".sign",color="pvalue") + facet_grid(.~.sign)
dev.off()
quit()
```

## map a zip file for sharing results

```sh
cd /home/jelbers/Downloads/heisenberg/DA0031

zip DA0031-2021_Sep_01-analysis1.zip \
2021_Sep_01/analysis.1/DA0031-2021_Sep_01-analysis.1-trimming_multiqc_report.html \
2021_Sep_01/analysis.1/dgea/gene-set-enrichment-all-animal-all-lateral.txt \
2021_Sep_01/analysis.1/dgea/all-animal-all-lateral/DA0031_2021_Sep_01_analysis.1.pca.pdf \
2021_Sep_01/analysis.1/dgea/all-animal-all-lateral/DESeq2_Animal-Lateral.csv \
2021_Sep_01/analysis.1/dgea/enhanced-volcano-all-animal-all-lateral.pdf \
2021_Sep_01/analysis.1/dgea/gene-set-enrichment-minus-animal.1-lateral.4.pdf \
2021_Sep_01/analysis.1/dgea/gene-set-enrichment-all-animal-all-lateral.pdf \
2021_Sep_01/analysis.1/id2sample.txt \
2021_Sep_01/analysis.1/dgea/enhanced-volcano-minus-animal.1-lateral.4.pdf \
2021_Sep_01/analysis.1/dgea/minus-animal.1-lateral.4/DA0031_2021_Sep_01_analysis.1.pca.pdf \
2021_Sep_01/analysis.1/dgea/minus-animal.1-lateral.4/DESeq2_Animal-Lateral.csv \
2021_Sep_01/analysis.1/dgea/gene-set-enrichment-minus-animal.1-lateral.4.txt
```
