# DA0023 - Single-cell RNA-sequencing of murine MADM labelled cortical cells

#### note that I have included some of the scripts in the README.md file, but they are also in the root directory of this repository
#### note2 I also use a mixture of relative and absolute paths sorry

## Initial meeting 8.6.2021:
single cell RNA sequencing 
goals are:
- comparing expression levels of wt and knockout cells of the same organism (six organisms analyzed)
- comparing expression levels in different cell types (classified according to expression patterns of certain genes) 
- check expression levels of transporter in wt cells


---
## Input files for test run

### Download Sequencing reads

    cd ~/novarino
    wget https://storage.googleapis.com/nih-sequence-read-archive/sra-src/SRR13012230/P1-P2_S1_L003_R1_001.fastq.gz.2
    wget https://storage.googleapis.com/nih-sequence-read-archive/sra-src/SRR13012230/P1-P2_S1_L003_R2_001.fastq.gz.2
    rename 's/.fastq.gz.2/.fastq.gz/' *.fastq.gz.2


### Modify sequencing reads
    
    mkdir -p ~/novarino/20210604_15_Mmu_Adult_corticalglial/data/HVC3MDSXY_4_R11191_20210302/demultiplexed/P1-P2
    module load seqtk/20210125
    # use seqtk to get only the first 28 bases (cell barcode of read 1)
    seqtk trimfq -e 122 P1-P2_S1_L003_R1_001.fastq.gz |pigz > ~/novarino/20210604_15_Mmu_Adult_corticalglial/data/HVC3MDSXY_4_R11191_20210302/demultiplexed/P1-P2/P1-P2_S1_L003_R1_001.fastq.gz &
    # just copy read 2 to the same directory
    cp P1-P2_S1_L003_R2_001.fastq.gz ~/novarino/20210604_15_Mmu_Adult_corticalglial/data/HVC3MDSXY_4_R11191_20210302/demultiplexed/P1-P2/P1-P2_S1_L003_R2_001.fastq.gz


### Get cellbarcodes for counting in STARsolo 2.7.9a Github commit# ac39348

    # for future reference as they are in a different place listed below
    wget https://github.com/10XGenomics/cellranger/raw/master/lib/python/cellranger/barcodes/3M-february-2018.txt.gz
    gunzip 3M-february-2018.txt.gz


### Get mouse genome and annotations

    # primary genome assembly
    wget http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M27/GRCm39.primary_assembly.genome.fa.gz
    # annotations for primary assembly
    wget http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M27/gencode.vM27.annotation.gtf.gz

    # decompress
    pigz -kd GRCm39.primary_assembly.genome.fa.gz
    pigz -kd gencode.vM27.annotation.gtf.gz


## Index reference genome with annotations for STARsolo

#### genomeGenerate.sh

    #!/bin/bash
    #SBATCH --job-name=genomeGenerate
    #SBATCH --output=STAR-2.7.9a.genomeGenerate.GRCm39.primary_assembly_gencode.vM27.annotation.%j.log
    #SBATCH --error=STAR-2.7.9a.genomeGenerate.GRCm39.primary_assembly_gencode.vM27.annotation.%j.err
    #SBATCH --nodes=1
    #SBATCH --cpus-per-task=48
    #SBATCH --time=3-24:00:00
    #SBATCH --mem=100G
    #SBATCH --export=NONE
    unset SLURM_EXPORT_ENV

    date
    echo 'log computing node'
    hostnamectl
    echo ''
    echo 'run star genomeGenerate'

    mkdir -p /nfs/scistore16/itgrp/jelbers/novarino/STARsolo/GRCm39.primary_assembly_gencode.vM27.annotation
    ~/bin/STAR/source/STAR \
    --limitGenomeGenerateRAM 100000000000 \
    --runThreadN 48 \
    --runMode genomeGenerate \
    --genomeDir /nfs/scistore16/itgrp/jelbers/novarino/STARsolo/GRCm39.primary_assembly_gencode.vM27.annotation \
    --genomeFastaFiles /nfs/scistore16/itgrp/jelbers/novarino/GRCm39.primary_assembly.genome.fa \
    --sjdbGTFfile /nfs/scistore16/itgrp/jelbers/novarino/gencode.vM27.annotation.gtf
    date


    sbatch genomeGenerate.sh


## Run STARsolo

#### test.sh

    #! /bin/bash
    set -ex

    # run STARsolo
    mkdir -p ${workingDir}/${projectDir}/cellranger
    cd ${workingDir}/${projectDir}/cellranger
    for s in ${sampleIDs}
    do
      echo '#!/bin/bash' > starsolo_${s}.slurm
      echo "#SBATCH --job-name=star_${s}" >> starsolo_${s}.slurm
      echo "#SBATCH --output=star_${s}_slurm.log" >> starsolo_${s}.slurm
      echo "#SBATCH --error=star_${s}_slurm.err" >> starsolo_${s}.slurm
      echo '#SBATCH --nodes=1' >> starsolo_${s}.slurm
      echo '#SBATCH --cpus-per-task=48' >> starsolo_${s}.slurm
      echo '#SBATCH --time=4:00:00' >> starsolo_${s}.slurm
      echo '#SBATCH --mem=100G' >> starsolo_${s}.slurm
      echo '#SBATCH --export=NONE' >> starsolo_${s}.slurm
      echo 'unset SLURM_EXPORT_ENV' >> starsolo_${s}.slurm

      echo "" >> starsolo_${s}.slurm
      echo "
        
        date
        echo 'log computing node'
        hostname
        echo ''
        echo 'run STARsolo'
        
        mkdir -p ${workingDir}/${projectDir}/cellranger/${s}
        cd ${workingDir}/${projectDir}/cellranger/${s}
        ~/bin/STAR/source/STAR \
        --outSAMtype BAM SortedByCoordinate \
        --readFilesCommand zcat \
        --soloUMIlen 12 \
        --runThreadN 48 \
        --genomeDir ${cellrangerIdx} \
        --readFilesIn ${workingDir}/${projectDir}/data/${requestID}/demultiplexed/${s}/${s}*R2*.fastq.gz ${workingDir}/${projectDir}/data/${requestID}/demultiplexed/${s}/${s}*R1*.fastq.gz \
        --soloType CB_UMI_Simple \
        --soloCBwhitelist /nfs/scistore16/itgrp/jelbers/20210209_15_Xla_Adult_spinalcordnuclei_combined/STARsolo/3M-february-2018.txt \
        --clipAdapterType CellRanger4 \
        --outFilterScoreMin 30 \
        --soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts \
        --soloUMIfiltering MultiGeneUMI_CR \
        --soloUMIdedup 1MM_CR \
        --soloFeatures Gene
        date
      " >> starsolo_${s}.slurm
      sbatch -W starsolo_${s}.slurm &
    done

    wait


### Variables to input to test.sh

    export workingDir="/nfs/scistore16/itgrp/jelbers/novarino" && \
    export projectDir="20210604_15_Mmu_Adult_corticalglial" && \
    export requestID="HVC3MDSXY_4_R11191_20210302" && \
    export sampleIDs="P1-P2" && \
    export cellrangerIdx="/nfs/scistore16/itgrp/jelbers/novarino/STARsolo/GRCm39.primary_assembly_gencode.vM27.annotation" && \
    export scriptDIR="/nfs/scistore12/itgrp/famman/Sandbox/DA0016" && \
    bash <(tail -n +3 test.sh |\
    sed '1i set -ex' |\
    sed '1i #! /bin/bash') >> ${workingDir}/${projectDir}/test.sh.log 2>&1 &


## Run seurat

#### seurat_for_STARsolo.sh

    #! /bin/bash
    set -ex

    ######
    mkdir -p ${workingDir}/${projectDir}/scripts
    cd ${workingDir}/${projectDir}/scripts
    cp ${scriptDIR}/{seurat_pre_process.R,functions.R} .
    sed 's/do.markers=T/do.markers=F/' functions.R > functions2.R

    # make directories
    mkdir -p ${workingDir}/${projectDir}/seurat
    cd ${workingDir}/${projectDir}/seurat


    # make samplestrings
    sampleIDstring=`echo ${sampleIDs} | tr " " "_"`

    echo '#!/bin/bash' > seurat.slurm
    echo "#SBATCH --job-name=seurat" >> seurat.slurm
    echo "#SBATCH --output=seurat_slurm.log" >> seurat.slurm
    echo "#SBATCH --error=seurat_slurm.err" >> seurat.slurm
    echo '#SBATCH --nodes=1' >> seurat.slurm
    echo '#SBATCH --cpus-per-task=2' >> seurat.slurm
    echo '#SBATCH --time=24:00:00' >> seurat.slurm
    echo '#SBATCH --mem=60G' >> seurat.slurm
    echo '#SBATCH --export=NONE' >> seurat.slurm
    echo 'unset SLURM_EXPORT_ENV' >> seurat.slurm
    echo "" >> seurat.slurm
    echo "
      module load R/4.0.3
      Rscript <(head -n 108 ${workingDir}/${projectDir}/scripts/seurat_pre_process.R |sed 's/outs\/filtered_feature_bc_matrix/Solo.out\/Gene\/filtered/' |sed 's/functions.R/functions2.R/') --path $workingDir --name $projectDir --aggr ${sampleIDstring} --species mmusculus
    " >> seurat.slurm
    sbatch -W seurat.slurm


### Variables to input to seurat_for_STARsolo.sh

    export workingDir="/nfs/scistore16/itgrp/jelbers/novarino" && \
    export projectDir="20210604_15_Mmu_Adult_corticalglial" && \
    export requestID="HVC3MDSXY_4_R11191_20210302" && \
    export sampleIDs="P1-P2" && \
    export cellrangerIdx="/nfs/scistore16/itgrp/jelbers/novarino/STARsolo/GRCm39.primary_assembly_gencode.vM27.annotation" && \
    export scriptDIR="/nfs/scistore12/itgrp/famman/Sandbox/DA0016" && \
    bash <(tail -n +3 seurat_for_STARsolo.sh |\
    sed '1i set -ex' |\
    sed '1i #! /bin/bash') >> ${workingDir}/${projectDir}/seurat_for_STARsolo.sh.log 2>&1 &


### Classify cells using clustifyr and the Mouse Cell Atlas data

    # use mamba or conda to install clustifyr-1.4.0 then upgrade it to clustifyr-1.5.0

    mamba create -n clustifyr-1.5.0 -c conda-forge -c bioconda bioconductor-clustifyr=1.4.0 r-seurat=4.0.2

    conda activate clustifyr-1.5.0

    # working directory containing the Seurat RDS saved file
    cd /nfs/scistore16/itgrp/jelbers/novarino/20210604_15_Mmu_Adult_corticalglial/seurat

    # this is the clustifyr upgrade
    wget https://github.com/rnabioco/clustifyr/archive/refs/tags/v1.5.0.tar.gz

    R
    install.packages("remotes")
    
    # you need clustifyrdata for the mouse cell atlas data
    remotes::install_github("rnabioco/clustifyrdata")

    install.packages("v1.5.0.tar.gz", repos = NULL, type="source")

    
    # load libraries in R

    library(clustifyr)
    library(clustifyrdata)
    library(ggplot2)
    library(cowplot)
    library(Seurat)

    # get the session information

    sessionInfo()

    #R version 4.1.0 (2021-05-18)
    #Platform: x86_64-conda-linux-gnu (64-bit)
    #Running under: Debian GNU/Linux 10 (buster)

    #Matrix products: default
    #BLAS/LAPACK: /nfs/scistore16/itgrp/jelbers/miniconda2/envs/clustifyr-1.5.0/lib/libopenblasp-r0.3.15.so

    #locale:
    # [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
    # [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
    # [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
    # [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
    # [9] LC_ADDRESS=C               LC_TELEPHONE=C            
    #[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

    #attached base packages:
    #[1] stats     graphics  grDevices utils     datasets  methods   base     

    #other attached packages:
    #[1] SeuratObject_4.0.1  Seurat_4.0.2        cowplot_1.1.1      
    #[4] ggplot2_3.3.3       clustifyrdata_1.1.0 clustifyr_1.5.0    

    #loaded via a namespace (and not attached):
    #  [1] fgsea_1.18.0                Rtsne_0.15                 
    #  [3] colorspace_2.0-1            deldir_0.2-10              
    #  [5] ellipsis_0.3.2              ggridges_0.5.3             
    #  [7] XVector_0.32.0              GenomicRanges_1.44.0       
    #  [9] spatstat.data_2.1-0         leiden_0.3.8               
    # [11] listenv_0.8.0               remotes_2.4.0              
    # [13] ggrepel_0.9.1               fansi_0.4.2                
    # [15] codetools_0.2-18            splines_4.1.0              
    # [17] polyclip_1.10-0             jsonlite_1.7.2             
    # [19] entropy_1.3.0               ica_1.0-2                  
    # [21] cluster_2.1.2               png_0.1-7                  
    # [23] uwot_0.1.10                 shiny_1.6.0                
    # [25] sctransform_0.3.2           spatstat.sparse_2.0-0      
    # [27] compiler_4.1.0              httr_1.4.2                 
    # [29] Matrix_1.3-4                fastmap_1.1.0              
    # [31] lazyeval_0.2.2              later_1.2.0                
    # [33] htmltools_0.5.1.1           tools_4.1.0                
    # [35] igraph_1.2.6                gtable_0.3.0               
    # [37] glue_1.4.2                  GenomeInfoDbData_1.2.6     
    # [39] RANN_2.6.1                  reshape2_1.4.4             
    # [41] dplyr_1.0.6                 fastmatch_1.1-0            
    # [43] Rcpp_1.0.6                  scattermore_0.7            
    # [45] Biobase_2.52.0              vctrs_0.3.8                
    # [47] nlme_3.1-152                lmtest_0.9-38              
    # [49] stringr_1.4.0               globals_0.14.0             
    # [51] mime_0.10                   miniUI_0.1.1.1             
    # [53] lifecycle_1.0.0             irlba_2.3.3                
    # [55] goftest_1.2-2               future_1.21.0              
    # [57] zlibbioc_1.38.0             MASS_7.3-54                
    # [59] zoo_1.8-9                   scales_1.1.1               
    # [61] spatstat.core_2.1-2         promises_1.2.0.1           
    # [63] MatrixGenerics_1.4.0        spatstat.utils_2.1-0       
    # [65] parallel_4.1.0              SummarizedExperiment_1.22.0
    # [67] RColorBrewer_1.1-2          SingleCellExperiment_1.14.1
    # [69] curl_4.3.1                  reticulate_1.20            
    # [71] pbapply_1.4-3               gridExtra_2.3              
    # [73] rpart_4.1-15                stringi_1.6.2              
    # [75] S4Vectors_0.30.0            BiocGenerics_0.38.0        
    # [77] BiocParallel_1.26.0         GenomeInfoDb_1.28.0        
    # [79] rlang_0.4.11                pkgconfig_2.0.3            
    # [81] matrixStats_0.59.0          bitops_1.0-7               
    # [83] lattice_0.20-44             tensor_1.5                 
    # [85] ROCR_1.0-11                 purrr_0.3.4                
    # [87] patchwork_1.1.1             htmlwidgets_1.5.3          
    # [89] tidyselect_1.1.1            parallelly_1.25.0          
    # [91] RcppAnnoy_0.0.18            plyr_1.8.6                 
    # [93] magrittr_2.0.1              R6_2.5.0                   
    # [95] IRanges_2.26.0              generics_0.1.0             
    # [97] DelayedArray_0.18.0         withr_2.4.2                
    # [99] mgcv_1.8-36                 pillar_1.6.1               
    #[101] fitdistrplus_1.1-5          abind_1.4-5                
    #[103] survival_3.2-11             RCurl_1.98-1.3             
    #[105] tibble_3.1.2                future.apply_1.7.0         
    #[107] crayon_1.4.1                KernSmooth_2.23-20         
    #[109] utf8_1.2.1                  spatstat.geom_2.1-0        
    #[111] plotly_4.9.3                grid_4.1.0                 
    #[113] data.table_1.14.0           digest_0.6.27              
    #[115] xtable_1.8-4                tidyr_1.1.3                
    #[117] httpuv_1.6.1                stats4_4.1.0               
    #[119] munsell_0.5.0               viridisLite_0.4.0


    # load the reference
    mca_ref = clustifyrdata::ref_MCA

    # load the processed Seurat file
    mca.matrix <- readRDS(file = "20210604_15_Mmu_Adult_corticalglial_processed.rds")

    # use clustifyr to correlate the counts to the mouse cell atlas cell types
    res <- clustify(
      input = mca.matrix, # matrix of normalized scRNA-seq counts (or SCE/Seurat object)
      cluster_col = "seurat_clusters", # name of column in meta.data containing cell clusters
      ref_mat = mca_ref,
      query = mca.matrix@assays$RNA@var.features,
      threshold=0.70
    )

    # plot the umap with ggplot2

    res2 <- data.frame("UMAP_1"=res$UMAP_1,
                       "UMAP_2"=res$UMAP_2,
                       "type"=as.factor(sub("\\(\\S+\\)","",res$type,perl=t)))
                       
    p<-  ggplot(data=res2, aes(x=UMAP_1, y=UMAP_2, color=type)) +
      geom_point(size=0.05) +
      xlab("UMAP_1") + ylab("UMAP_2") +
      theme_minimal() +
      theme(axis.line.x.bottom = element_line(),
            axis.line.y.left = element_line(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()) +
      guides(colour = guide_legend(override.aes = list(size=3)))
    ggsave("clustifyr.png", p, height=6, width=6,dpi=300)



    res <- clustify(
      input = mca.matrix, # matrix of normalized scRNA-seq counts (or SCE/Seurat object)
      cluster_col = "seurat_clusters", # name of column in meta.data containing cell clusters
      ref_mat = mca_ref,
      query = mca.matrix@assays$RNA@var.features,
      dr = "umap"
    )

    res2 <- data.frame("UMAP_1"=res$UMAP_1,
                       "UMAP_2"=res$UMAP_2,
                       "type"=as.factor(res$type))
    p<-  ggplot(data=res2, aes(x=UMAP_1, y=UMAP_2, color=type)) +
      geom_point() +
      xlab("UMAP_1") + ylab("UMAP_2") +
      theme_bw()
    ggsave("clustifyr.png", p, height=6, width=6,dpi=300)
    
#### Example cell classification with clustifyr
![clustifyr.png](clustifyr.png?raw=true "Example cell classification with clustifyr")
